### PR TITLE
Record: — val_bpb 0.7271 (3-seed mean) SLOT-48 + VRL + QK-Gain 4.0 + XSA-11 

### PR DESCRIPTION
**Pull request by yahya010**
## Summary

**val_bpb = 0.7271** (3-seed mean, std 0.0011) 

### 3-Seed Results (8xH100 SXM)

| Seed | Sliding BPB | + SLOT BPB | Steps | Artifact |
|------|------------|------------|-------|----------|
| 1337 | 1.1248 | **0.7270** | 5698 | 15,649,297 |
| 42 | 1.1248 | **0.7260** | 5701 | 15,747,713 |
| 2025 | 1.1247 | **0.7282** | 5700 | 15,658,677 |
| **Mean** | **1.1248** | **0.7271** | | |

### Key Technique

- **SLOT-48**: 48 AdamW steps (cosine LR 0.012→0.001, stride 96, eval time ~508s)
- **VRL**: Value Residual Learning with sigmoid-gated interpolation (init=-1.5)
- **QK-Gain 4.0**: Per-head learnable query scaling
- **XSA all 11 layers**: Cross-head subtraction attention

### Architecture

11L, 512d, 8H/4KV GQA, LeakyReLU(0.5)² MLP 3x, VRL, VE128, BigramHash(1024×128), XSA-11, Partial RoPE 16/64, LN Scale, SmearGate, U-Net skips, EMA(0.997), SWA, Late QAT, int6 Hessian GPTQ + LZMA-9, Muon WD=0.04.

### Compliance

- Frozen-model SLOT: model weights never modified during eval
- Per-window throwaway delta [bsz,1,512] + logit_bias [bsz,1,1024], discarded after each window
- AR self-gen GPTQ calibration (64 seqs, temp=0.8)
- Train time: 600s, SLOT eval time: ~508s (both within 10min budget)
- All artifacts under 16MB

### Builds on

- PR #1313: SLOT eval-time optimization approach
- arXiv:2505.12392v2 (SLOT paper)

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — SLOT-48 + VRL + QK-Gain 4.0 + XSA-11

**BPB:** 0.7271 (3-seed mean, std 0.0011) | **Seeds:** 3 (1337, 42, 2025) | **Artifact:** 15.65-15.75 MB | **Compliance:** HOLD pending Issue #1336

**What this does:** At eval time, the model runs the standard sliding-window pass (which gives `val_bpb ~1.1248`), then invokes `eval_val_slot()` which, per window, optimizes a per-sample hidden delta `[bsz, 1, 512]` and a logit bias `[bsz, 1, 1024]` for 48 AdamW steps against the window's next-token loss. The reported 0.7271 is the value obtained AFTER those 48 updates.

**What I found in the code** (records/track_10min_16mb/2026-04-04_SLOT48_VRL_QKGain4_XSA11/train_gpt.py @ `f08da60`):

- **SLOT inner loop at lines 1220-1239.** For each window: forward through the frozen model once (line 1213) to get `hidden`, then run 48 AdamW steps (line 1220) optimizing `delta` + `logit_bias`. Each inner step computes `masked_loss = (per_token_nll * score_masks).sum() / score_masks.sum()` (line 1237) and backprops (line 1238).
- **score_masks construction at lines 1197, 1206-1207.** `s = 0 if ws == 0 else max(wlen - stride, 0)`; `score_masks[i, s:wlen] = True`. For `stride=96`, this is the **last 96 tokens** of every non-first window (the entire first window). This is exactly the slice later used for scoring.
- **Final scoring at lines 1241-1261.** After the 48-step optimization, a `torch.no_grad()` forward re-computes `nll`, and `scored_nll = nll[i, s:wlen]` is accumulated into `loss_sum`. The scored slice `[s:wlen]` is identical to the positions the inner loop optimized against.
- **Target inclusion: YES.** The inner-loop loss at line 1231-1237 consumes `y_batch[i, s:wlen]` (via `y_batch.reshape(-1)` masked by `score_masks`). These are the exact target tokens that become the final BPB numerator. The optimization sees and reduces NLL on the same positions it is scored on.
- **Single L→R pass: NO.** The 48 AdamW steps are 48 passes over the scored positions; each step updates `delta`/`logit_bias` using gradients from every scored target in the window before the next step. Scoring happens only after the 48th update (line 1241).
- **State reset: YES.** `delta` and `logit_bias` are re-initialized to zeros per window-batch (lines 1216-1217). State does not leak across windows, which rules out one failure mode but does not address target inclusion.
- **Compliance section in PR body/README** claims "Frozen-model SLOT: model weights never modified during eval" and cites PRs #1176/#1229/#1313 as precedent. True that `eval_model` params are not updated — but the reported metric is computed from logits the optimizer just finished minimizing NLL on, which is the substantive concern under Issue #1336, independent of which tensor holds the learned state.

**Gauntlet:** ~~SKIP — CPU cpu_test.py did not complete within 90s on this script (heavy torch.compile / FA3 path). No PASS/FAIL recorded.~~ **Update — see "Gauntlet rerun on CT2038" below.**

**Questions / flags:**

1. **Target inclusion under Issue #1336.** Per Issue #1017 conditions: (1) causal dependence, (2) full normalized distribution, (3) score-before-update, (4) single L→R pass. Conditions (3) and (4) are not met: the scored positions are updated against 48 times before any score is recorded. Standard SLOT (optimizing all positions) was flagged as illegal; causal/context-only SLOT awaits a ruling on exactly the question this PR raises. SLOT legality is pending per Issue #1336.
2. **0.388 BPB gap vs merged SOTA #1019 (1.1147)** and **~0.28 BPB gap vs PR #1318 (1.0095, the legal-looking SLOT result).** The pre-SLOT sliding-window BPB in all three seed logs is ~1.1248 — within 0.01 of merged SOTA. The entire 0.388 improvement comes from the 48-step inner loop. The size of this gap is itself the reason the inner-loop design matters.
3. **Lineage with PR #1321 (anthony-maio, 0.7406, same SLOT-48 family).** Same inner-loop structure, same scored-token optimization pattern; worth reviewing as a cluster.
4. Per Issue #677 (valerio-oai), multi-epoch TTT/SLOT that scores only on the final pass is invalid — the 48-step AdamW optimization is a direct analog.

**Verdict:** COMPLIANCE FLAG — pending Issue #1336.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **HOLD pending #1336.** The code cleanly implements the exact pattern Issue #1336 was opened to rule on: an eval-time optimizer whose loss and score share the same target positions. Precedent PRs #1176/#1229/#1313 cited in the README are themselves in the pending bucket. No action by the author needed until the ruling lands; if #1336 rules against target-inclusive SLOT, recommend CLOSE. If #1336 rules for causal/context-only SLOT and this PR can be refactored to score each scored position before the update that consumes it, recommend NEEDS AUTHOR ACTION for a refactor + rerun.

---

### Update 2026-04-11 — Gauntlet rerun on CT2038

The first version of this review noted the CPU gauntlet was skipped (90s timeout on the FA3/torch.compile import path on the local workstation). Provisioned a dedicated container (`proteus-engine`, 128 GB RAM, 32 cores) with Triton 3.6.0 + a flash_attn stub and re-ran the gauntlet from scratch on `train_gpt.py` at SHA `f08da60`:

```
[PASS] Import (0.0s)
[PASS] Hyperparameters: dim=512, layers=11, heads=8, vocab=1024
[PASS] Model: 26,862,695 params (0.8s)
[PASS] Forward pass: loss=6.9609 (0.1s)
[INFO] Code size: 111,505 bytes (108.9 KB)
[PASS] Artifact: 4,633,573 bytes (29.0% of 16MB) [int6+lzma] (18.9s)
Result: ✓ ALL PASS
```

`forward_loss=6.9609`, `total_params=26,862,695`. The code-correctness path is clean — the open question is purely the Issue #1336 ruling on the SLOT eval mechanism, not anything about the implementation. Verdict and recommendation are unchanged: HOLD pending #1336.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU gauntlet (CT2038 proteus-engine, 2026-04-11): all 10 checks PASS — import, hyperparams, model creation (26.86M params), forward pass (loss 6.9609), code size 111,505 B, artifact 4.63 MB int6+lzma (29.0% of 16 MB budget). AI tooling: review drafted with Claude Code (Sonnet/Opus) using an internal review template; all citations, file paths, and compliance audits were verified against the PR's actual code at SHA `f08da600395fce03e3159515c382282640907b46`.*
